cmd/history_index.cpp

```diff
@@ -0,0 +1,189 @@
+    app.add_flag("--storage", storage, "Do history of storages");
+
+    CLI11_PARSE(app, argc, argv);
+
+    // Check data.mdb exists in provided directory
+    boost::filesystem::path db_file{boost::filesystem::path(db_path) / boost::filesystem::path("data.mdb")};
+    if (!boost::filesystem::exists(db_file)) {
+        SILKWORM_LOG(LogError) << "Can't find a valid TG data file in " << db_path << std::endl;
+        return -1;
+    }
+    fs::path datadir(db_path);
+    fs::path etl_path(datadir.parent_path() / fs::path("etl-temp"));
+    fs::create_directories(etl_path);
+    etl::Collector collector(etl_path.string().c_str(), /* flush size */ 512 * kMebi);
+
+    lmdb::DatabaseConfig db_config{db_path};
+    db_config.set_readonly(false);
+    std::shared_ptr<lmdb::Environment> env{lmdb::get_env(db_config)};
+    std::unique_ptr<lmdb::Transaction> txn{env->begin_rw_transaction()};
+    // We take data from header table and transform it and put it in blockhashes table
+    lmdb::TableConfig changeset_config =
+        storage ? db::table::kPlainStorageChangeSet : db::table::kPlainAccountChangeSet;
+    lmdb::TableConfig index_config = storage ? db::table::kStorageHistory : db::table::kAccountHistory;
+    const char *stage_key = storage ? db::stages::kStorageHistoryIndexKey : db::stages::kAccountHistoryKey;
+    auto changeset_table{txn->open(changeset_config)};
+    std::unordered_map<std::string, roaring::Roaring64Map> bitmaps;
+
+    try {
+        auto last_processed_block_number{db::stages::get_stage_progress(*txn, stage_key)};
+
+        if (full) {
+            last_processed_block_number = 0;
+            txn->open(index_config, MDB_CREATE)->clear();
+        }
+
+        // Extract
+        Bytes start(8, '\0');
+        boost::endian::store_big_u64(&start[0], last_processed_block_number);
+        MDB_val mdb_key{db::to_mdb_val(start)};
+        MDB_val mdb_data;
+
+        SILKWORM_LOG(LogInfo) << "Started " << (storage ? "Storage" : "Account") << " Index Extraction" << std::endl;
+
+        size_t allocated_space{0};
+        uint64_t block_number{0};
+        int rc{changeset_table->seek(&mdb_key, &mdb_data)};  // Sets cursor to nearest key greater equal than this
+        while (!rc) {                                        /* Loop as long as we have no errors*/
+            std::string composite_key;
+            if (storage) {
+                char composite_key_array[kHashLength + kAddressLength];
+                std::memcpy(&composite_key_array[0], &static_cast<uint8_t *>(mdb_key.mv_data)[8], kAddressLength);
+                std::memcpy(&composite_key_array[kAddressLength], static_cast<uint8_t *>(mdb_data.mv_data),
+                            kHashLength);
+                composite_key = std::string(composite_key_array);
+            } else {
+                composite_key = std::string(static_cast<char *>(mdb_data.mv_data), kAddressLength);
+            }
```

Comment:
I'd evaluate if there are benefits in performance creating two variants of a lambda function in charge of bulding the key on behalf of `storage`. It should help removing an `if` branch for every block loop.
I agree this is not the heavy load bit.